src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/SignedCms.cs

```diff
@@ -152,6 +170,59 @@ internal void Decode(ReadOnlyMemory<byte> encodedMessage)
             _hasData = true;
         }
 
+        internal static ReadOnlyMemory<byte> GetContent(
+            ReadOnlyMemory<byte> wrappedContent,
+            string contentType)
+        {
+            // Read the input.
+            //
+            // PKCS7's id-data is written in both PKCS#7 and CMS as an OCTET STRING wrapping
+            // the arbitrary bytes, so the OCTET STRING must always be present.
+            //
+            // For other types, CMS says to always write an OCTET STRING, and to put the properly
+            // encoded data within it.
+            // PKCS#7 originally ommitted the OCTET STRING wrapper for this model, so this is the
+            // dynamic adapter.
+            //
+            // See https://tools.ietf.org/html/rfc5652#section-5.2.1
+            byte[] rented = null;
+            int bytesWritten = 0;
+            try
+            {
+                AsnReader reader = new AsnReader(wrappedContent, AsnEncodingRules.BER);
+
+                if (reader.TryGetPrimitiveOctetStringBytes(out ReadOnlyMemory<byte> inner))
+                {
+                    return inner;
+                }
+
+                rented = ArrayPool<byte>.Shared.Rent(wrappedContent.Length);
+
+                if (!reader.TryCopyOctetStringBytes(rented, out bytesWritten))
```

Comment:
Out of curiosity: When is TryGetPrimitiveOctetStringBytes going to fail and TryCopyOctetStringBytes will succeed?

Reply:
Is it the primitive vs non-primitive?

Reply:
Yep.  TryGetPrimitive returns false if the tag is correct, but the Primitive vs Constructed is not.  TryCopy can handle both.